app/controllers/oidc_users_controller.rb

```diff
@@ -42,6 +42,14 @@ def update
     render json: attributes.merge(sub: user.sub)
   end
 
+  def destroy
+    user = OidcUser.find_by!(sub: params.fetch(:subject_identifier))
+    user.destroy!
+    head :no_content and return
+  rescue ActiveRecord::RecordNotFound
+    head :not_found and return
```

Comment:
I'm unsure if this is needed or not.
On the one hand we're doing a search here, and if we don't rescue then we'd return a `500` instead of a `404` which seems less descriptive to me.

But on the other, I'm not sure if tis is a code path that should ever be reachable.
A user should not be able to submit a subject identifier other than their own... which should be present...

If folks think this is OTT, can remove it.

Reply:
> A user should not be able to submit a subject identifier other than their own..
True, _but_ that doesn't guarantee that it's present at the time the request is processed. There may be multiple requests in parallel, or async job retries, or someone doing an operation in the console just at the time the users request makes it through the queue, and so on.

That being said, Rails should rescue any `ActiveRecord::NotFound` error in a controller with a 404 by default (see the note about `config.action_dispatch.rescue_responses` in [Configuring Rails Apps](https://guides.rubyonrails.org/configuring.html#configuring-action-dispatch). It looks like we haven't done anything to alter this default -  from the console, run:
```
> Rails.configuration.action_dispatch.rescue_responses
=> {"ActiveRecord::RecordNotFound"=>:not_found,
 "ActiveRecord::StaleObjectError"=>:conflict,
 "ActiveRecord::RecordInvalid"=>:unprocessable_entity,
 "ActiveRecord::RecordNotSaved"=>:unprocessable_entity}
```

so I think we should be ok without this rescue clause. You could always add a test case for this?

Reply:
Right ok yeah so too broad in the rescue, that makes sense. I'll simplify this!

Reply:
🤔 humm so when I remove this i get:
```
Failures:                                                                                                                                                                                                          
                                                                                                                                                                                                                   
  1) OIDC Users endpoint DELETE when a user does not exist does not change the count of users and returns not found                                                                                                
     Failure/Error: user = OidcUser.find_by!(sub: params.fetch(:subject_identifier))                                                                                                                               
                                                                                                                                                                                                                   
     ActiveRecord::RecordNotFound:                                                                                                                                                                                 
       Couldn't find OidcUser                                                                                                                                                                                      
     # ./app/controllers/oidc_users_controller.rb:46:in `destroy'                                                                                                                                                  
     # ./spec/requests/oidc_users_spec.rb:103:in `block (5 levels) in <top (required)>'                                                                                                                            
     # ./spec/requests/oidc_users_spec.rb:103:in `block (4 levels) in <top (required)>'     
  ```
[Triggered by this line of the tests](https://github.com/alphagov/account-api/pull/154/commits/3e90b2f995b4775e41ed87341a3b1cfe926016ed#diff-17d8a2354955c9bf68280e255dd8117c4c5baa2d51037d9b692390e6e6b9b354R103)

I tried removing the change test and just checking for the response code, but the tests still freak out at the existance of the error. And i tried `be_not_found` on the block (on the off chance the block syntax would let it bubble up to whaterver returned 404)... but no joy `be_not_found` does not support rspec block syntax.

When does action dispatch kick in here?
Id expected to be able to test this by just removing the clause and having my existing 404 test work... but perhaps not with request specs?

Reply:
Ah, request specs are not 'true' full-stack tests and are perhaps closer to controller specs in this respect. It's Rails trying to be helpful and intercepting the error to show you the full details in test mode, rather than giving a true reflection of what the user would see in production. See [this answer when this was raised as an issue in rspec-rails](https://github.com/rspec/rspec-rails/issues/2024#issuecomment-420809675).

There's two options, really:
1. test this with a feature spec instead, or
2. add a [workaround that tells Rails not to intercept the error](https://github.com/rspec/rspec-rails/issues/2024#issuecomment-420646336)

Reply:
I suppose if we can trust that the interceptor works then the test could be:

```
expect { delete oidc_user_path(subject_identifier: subject_identifier), params: params, headers: headers }.to raise(ActiveRecord::RecordNotFound)
```

Then i can add a follow up feature test to check we get a 404?

Reply:
[This is a great link](https://github.com/rspec/rspec-rails/issues/2024#issuecomment-420646336) by the way, thanks very much.

Super interetsing thread and to see an example of how-to!
I'm somewhat leaning that way now...

Reply:
It works! I really like this,
It's explicit and contained. I'll note the reasons and link to this thread in a commit then force push.

Thanks Al! Great review